File: strands-ts/src/tools/executors/executor.ts
import { normalizeError } from '../../errors.js'
import { AfterToolCallEvent, BeforeToolCallEvent, ToolStreamUpdateEvent } from '../../hooks/events.js'
import { InterruptError, interruptFromAgent } from '../../interrupt.js'
import { createMiddlewareInterrupt } from '../../middleware/interrupt.js'
import { ExecuteToolStage } from '../../middleware/index.js'
import { deepCopy } from '../../types/json.js'
import { TextBlock, ToolResultBlock } from '../../types/messages.js'

import type { Agent } from '../../agent/agent.js'
import type { ToolUseData } from '../../hooks/events.js'
import type { ExecuteToolContext, ExecuteToolResult, MiddlewareRegistry } from '../../middleware/index.js'
import type { Meter } from '../../telemetry/meter.js'
import type { Tracer } from '../../telemetry/tracer.js'
import type { AgentStreamEvent, InvocationState } from '../../types/agent.js'
import type { InterruptParams } from '../../types/interrupt.js'
import type { JSONValue } from '../../types/json.js'
import type { Message, ToolResultBlockData, ToolUseBlock } from '../../types/messages.js'
import type { Tool, ToolContext } from '../tool.js'

/**
 * Dependencies supplied for one tool-executor invocation.
 *
 * @internal
 */
export interface ToolExecutorOptions {
  /** Agent whose tools are being executed. */
  readonly agent: Agent
  /** Registry containing middleware for the agent. */
  readonly middlewareRegistry: MiddlewareRegistry
  /** Tracer used to record tool calls. */
  readonly tracer: Tracer
  /** Meter used to record tool-call metrics. */
  readonly meter: Meter
}

/**
 * Input for executing the tool calls from one model turn.
 *
 * @internal
 */
export interface ToolExecutionInput {
  /** Tool calls to execute. */
  readonly toolUseBlocks: readonly ToolUseBlock[]
  /** Accumulates completed tool results and may remain partial when execution exits early. */
  readonly toolResultBlocks: ToolResultBlock[]
  /** State shared across the current agent invocation. */
  readonly invocationState: InvocationState
  /** Assistant message that requested the tool calls. */
  readonly assistantMessage: Message
  /** Results restored when resuming interrupted execution. */
  readonly completedToolResults?: ReadonlyMap<string, ToolResultBlock>
}

/**
 * Shared pipeline for executing tool calls.
 *
 * @internal
 */
export abstract class ToolExecutor {
  /**
   * Executes the tool calls from one model turn.
   *
   * @param options - Agent dependencies used to execute tools
   * @param input - Tool calls and invocation state
   * @returns Stream of tool lifecycle events
   * @internal
   */
  abstract execute(
    options: ToolExecutorOptions,
    input: ToolExecutionInput
  ): AsyncGenerator<AgentStreamEvent, void, undefined>

  // Tool lookup and tool-body failures become ToolResultBlocks so the model can
  // respond to them; lifecycle and middleware failures may still propagate.
  protected async *executeTool(
    options: ToolExecutorOptions,
    toolUseBlock: ToolUseBlock,
    invocationState: InvocationState
  ): AsyncGenerator<AgentStreamEvent, ToolResultBlock, undefined> {
    const registryTool = options.agent.toolRegistry.get(toolUseBlock.name)

    // Callbacks may replace or mutate this value inside BeforeToolCallEvent.
    let toolUse = {
      name: toolUseBlock.name,
      toolUseId: toolUseBlock.toolUseId,
      input: toolUseBlock.input,
    }

    while (true) {
      const beforeToolCallEvent = new BeforeToolCallEvent({
        agent: options.agent,
        toolUse,
        tool: registryTool,
        invocationState,
      })
      yield beforeToolCallEvent

      toolUse = {
        ...beforeToolCallEvent.toolUse,
        toolUseId: toolUseBlock.toolUseId,
      }
      // selectedTool takes precedence over resolving a hook-renamed toolUse;
      // otherwise retain the original registry lookup. Resolve before cancellation
      // so AfterToolCallEvent reports the same effective tool on every path.
      const effectiveTool =
        beforeToolCallEvent.selectedTool ??
        (toolUse.name !== toolUseBlock.name ? options.agent.toolRegistry.get(toolUse.name) : registryTool)

      if (beforeToolCallEvent.cancel) {
        const cancelMessage =
          typeof beforeToolCallEvent.cancel === 'string' ? beforeToolCallEvent.cancel : 'Tool cancelled by hook'
        const cancelResult = new ToolResultBlock({
          toolUseId: toolUse.toolUseId,
          status: 'error',
          content: [new TextBlock(cancelMessage)],
        })
        const afterToolCallEvent = new AfterToolCallEvent({
          agent: options.agent,
          toolUse,
          tool: effectiveTool,
          result: cancelResult,
          invocationState,
        })
        yield afterToolCallEvent
        if (this._shouldRetry(options.agent, afterToolCallEvent)) {
          continue
        }
        return this._normalizeToolResultId(afterToolCallEvent.result, toolUseBlock.toolUseId)
      }

      const toolResult = this._normalizeToolResultId(
        yield* this._executeToolWithMiddleware(options, effectiveTool, toolUse, invocationState),
        toolUseBlock.toolUseId
      )
      const error = toolResult.error
      const afterToolCallEvent = new AfterToolCallEvent({
        agent: options.agent,
        toolUse,
        tool: effectiveTool,
        result: toolResult,
        invocationState,
        ...(error !== undefined && { error }),
      })
      yield afterToolCallEvent

      if (this._shouldRetry(options.agent, afterToolCallEvent)) {
        continue
      }

      // Return the hook-transformed result so downstream events and the model
      // observe the same value.
      return this._normalizeToolResultId(afterToolCallEvent.result, toolUseBlock.toolUseId)
    }
  }

  private _shouldRetry(agent: Agent, afterToolCallEvent: AfterToolCallEvent): boolean {
    return afterToolCallEvent.retry === true && !agent.cancelSignal.aborted
  }

  private _normalizeToolResultId(result: ToolResultBlock, toolUseId: string): ToolResultBlock {
    if (result.toolUseId === toolUseId) {
      return result
    }

    return new ToolResultBlock({
      toolUseId,
      status: result.status,
      content: result.content,
      ...(result.error !== undefined && { error: result.error }),
    })
  }

  // Keys and serialized results use model-issued toolUseIds so resume and
  // provider correlation match the assistant's tool-use blocks.
  protected _storePendingToolExecution(
    options: ToolExecutorOptions,
    assistantMessage: Message,
    completedToolResults: ReadonlyMap<string, ToolResultBlock>
  ): void {
    const serializedResults: Record<string, { toolResult: ToolResultBlockData }> = {}
    for (const [toolUseId, result] of completedToolResults) {
      serializedResults[toolUseId] = result.toJSON()
    }
    options.agent._interruptState.setPendingToolExecution({
      assistantMessageData: assistantMessage.toJSON(),
      completedToolResults: serializedResults,
    })
  }

  private async *_executeToolWithMiddleware(
    options: ToolExecutorOptions,
    tool: Tool | undefined,
    toolUse: ToolUseData,
    invocationState: InvocationState
  ): AsyncGenerator<AgentStreamEvent, ToolResultBlock, undefined> {
    const context: ExecuteToolContext = {
      agent: options.agent,
      tool,
      toolUse: deepCopy(toolUse) as unknown as ToolUseData,
      invocationState,
      interrupt: createMiddlewareInterrupt(
        options.agent._interruptState,
        `middleware:executeTool:${toolUse.toolUseId}`
      ),
    }

    // async function* does not bind lexical `this`; capture the executor for the terminal callback.
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const executor = this
    const middlewareResult = yield* options.middlewareRegistry.invoke(
      ExecuteToolStage,
      context,
      async function* (
        middlewareContext: ExecuteToolContext
      ): AsyncGenerator<AgentStreamEvent, ExecuteToolResult, undefined> {
        const result = yield* executor._executeToolCore(
          options,
          middlewareContext.tool,
          middlewareContext.toolUse,
          middlewareContext.invocationState
        )
        return { result }
      }
    )
    return middlewareResult.result
  }

  private async *_executeToolCore(
    options: ToolExecutorOptions,
    effectiveTool: Tool | undefined,
    toolUse: ToolUseData,
    invocationState: InvocationState
  ): AsyncGenerator<AgentStreamEvent, ToolResultBlock, undefined> {
    const toolSpan = options.tracer.startToolCallSpan({ tool: toolUse })
    const toolStartTime = Date.now()

    let toolResult: ToolResultBlock | undefined
    let error: Error | undefined

    try {
      if (!effectiveTool) {
        toolResult = new ToolResultBlock({
          toolUseId: toolUse.toolUseId,
          status: 'error',
          content: [new TextBlock(`Tool '${toolUse.name}' not found in registry`)],
        })
      } else {
        const toolContext: ToolContext = {
          toolUse: {
            name: toolUse.name,
            toolUseId: toolUse.toolUseId,
            input: toolUse.input,
          },
          agent: options.agent,
          invocationState,
          interrupt: <T = JSONValue>(params: InterruptParams): T =>
            interruptFromAgent<T>(options.agent, `tool:${toolUse.toolUseId}:${params.name}`, params, 'tool'),
        }

        // Iterate manually to wrap raw tool events at the agent boundary and
        // re-enter the tool span for every asynchronous step.
        const toolGenerator = options.tracer.withSpanContext(toolSpan, () => effectiveTool.stream(toolContext))
        let toolNext = await options.tracer.withSpanContext(toolSpan, () => toolGenerator.next())
        while (!toolNext.done) {
          yield new ToolStreamUpdateEvent({ agent: options.agent, event: toolNext.value, invocationState })
          toolNext = await options.tracer.withSpanContext(toolSpan, () => toolGenerator.next())
        }

        toolResult =
          toolNext.value ??
          new ToolResultBlock({
            toolUseId: toolUse.toolUseId,
            status: 'error',
            content: [new TextBlock(`Tool '${toolUse.name}' did not return a result`)],
          })
        error = toolNext.value?.error
      }

      return toolResult
    } catch (caughtError) {
      if (caughtError instanceof InterruptError) {
        throw caughtError
      }

      error = normalizeError(caughtError)
      toolResult = new ToolResultBlock({
        toolUseId: toolUse.toolUseId,
        status: 'error',
        content: [new TextBlock(error.message)],
        error,
      })
      return toolResult
    } finally {
      // Record the raw execution outcome before AfterToolCallEvent can transform it.
      options.tracer.endToolCallSpan(toolSpan, {
        ...(toolResult !== undefined && { toolResult }),
        ...(error !== undefined && { error }),
      })
      options.meter.endToolCall({
        tool: toolUse,
        duration: Date.now() - toolStartTime,
        success: toolResult?.status === 'success',
      })
    }
  }
}

File: strands-py/src/strands/tools/executors/_executor.py
"""Abstract base class for tool executors.

Tool executors are responsible for determining how tools are executed (e.g., concurrently, sequentially, with custom
thread pools, etc.).
"""

import abc
import logging
import time
from collections.abc import AsyncGenerator
from typing import TYPE_CHECKING, Any, cast

from opentelemetry import trace as trace_api

from ..._middleware.stages import ExecuteToolContext, ExecuteToolStage
from ...experimental.hooks.events import BidiAfterToolCallEvent, BidiBeforeToolCallEvent
from ...hooks import AfterToolCallEvent, BeforeToolCallEvent
from ...interrupt import InterruptException
from ...telemetry.metrics import Trace
from ...telemetry.tracer import get_tracer, serialize
from ...types._events import ToolCancelEvent, ToolInterruptEvent, ToolResultEvent, ToolStreamEvent, TypedEvent
from ...types.content import Message
from ...types.interrupt import Interrupt
from ...types.tools import ToolChoice, ToolChoiceAuto, ToolConfig, ToolResult, ToolUse
from ..structured_output._structured_output_context import StructuredOutputContext

if TYPE_CHECKING:  # pragma: no cover
    from ...agent import Agent
    from ...experimental.bidi import BidiAgent

logger = logging.getLogger(__name__)


class ToolExecutor(abc.ABC):
    """Abstract base class for tool executors."""

    @staticmethod
    def _is_agent(agent: "Agent | BidiAgent") -> bool:
        """Check if the agent is an Agent instance, otherwise we assume BidiAgent.

        Note, we use a runtime import to avoid a circular dependency error.
        """
        from ...agent import Agent

        return isinstance(agent, Agent)

    @staticmethod
    async def _invoke_before_tool_call_hook(
        agent: "Agent | BidiAgent",
        tool_func: Any,
        tool_use: ToolUse,
        invocation_state: dict[str, Any],
    ) -> tuple[BeforeToolCallEvent | BidiBeforeToolCallEvent, list[Interrupt]]:
        """Invoke the appropriate before tool call hook based on agent type."""
        kwargs = {
            "selected_tool": tool_func,
            "tool_use": tool_use,
            "invocation_state": invocation_state,
        }
        event = (
            BeforeToolCallEvent(agent=cast("Agent", agent), **kwargs)
            if ToolExecutor._is_agent(agent)
            else BidiBeforeToolCallEvent(agent=cast("BidiAgent", agent), **kwargs)
        )

        return await agent.hooks.invoke_callbacks_async(event)

    @staticmethod
    async def _invoke_after_tool_call_hook(
        agent: "Agent | BidiAgent",
        selected_tool: Any,
        tool_use: ToolUse,
        invocation_state: dict[str, Any],
        result: ToolResult,
        exception: Exception | None = None,
        cancel_message: str | None = None,
    ) -> tuple[AfterToolCallEvent | BidiAfterToolCallEvent, list[Interrupt]]:
        """Invoke the appropriate after tool call hook based on agent type."""
        kwargs = {
            "selected_tool": selected_tool,
            "tool_use": tool_use,
            "invocation_state": invocation_state,
            "result": result,
            "exception": exception,
            "cancel_message": cancel_message,
        }
        event = (
            AfterToolCallEvent(agent=cast("Agent", agent), **kwargs)
            if ToolExecutor._is_agent(agent)
            else BidiAfterToolCallEvent(agent=cast("BidiAgent", agent), **kwargs)
        )

        return await agent.hooks.invoke_callbacks_async(event)

    @staticmethod
    def _should_retry(agent: "Agent | BidiAgent", after_event: AfterToolCallEvent | BidiAfterToolCallEvent) -> bool:
        """Return whether a hook-requested retry should run.

        Cancellation is terminal: retrying a locally cancelled tool can spin indefinitely
        while delaying the caller's requested agent stop.
        """
        if not getattr(after_event, "retry", False):
            return False
        if cast(dict[str, Any], after_event.result).get("cancelled") is True:
            return False
        return not (ToolExecutor._is_agent(agent) and agent._cancel_signal.is_set())

    @staticmethod
    async def _stream(
        agent: "Agent | BidiAgent",
        tool_use: ToolUse,
        tool_results: list[ToolResult],
        invocation_state: dict[str, Any],
        structured_output_context: StructuredOutputContext | None = None,
        **kwargs: Any,
    ) -> AsyncGenerator[TypedEvent, None]:
        """Stream tool events.

        This method adds additional logic to the stream invocation including:

        - Tool lookup and validation
        - Before/after hook execution
        - Tracing and metrics collection
        - Error handling and recovery
        - Interrupt handling for human-in-the-loop workflows

        Args:
            agent: The agent (Agent or BidiAgent) for which the tool is being executed.
            tool_use: Metadata and inputs for the tool to be executed.
            tool_results: List of tool results from each tool execution.
            invocation_state: Context for the tool invocation.
            structured_output_context: Context for structured output management.
            **kwargs: Additional keyword arguments for future extensibility.

        Yields:
            Tool events with the last being the tool result.
        """
        logger.debug("tool_use=<%s> | streaming", tool_use)
        tool_name = tool_use["name"]
        structured_output_context = structured_output_context or StructuredOutputContext()

        tool_info = agent.tool_registry.dynamic_tools.get(tool_name)
        tool_func = tool_info if tool_info is not None else agent.tool_registry.registry.get(tool_name)
        tool_spec = tool_func.tool_spec if tool_func is not None else None

        current_span = trace_api.get_current_span()
        if current_span and tool_spec is not None:
            current_span.set_attribute("gen_ai.tool.description", tool_spec["description"])
            input_schema = tool_spec["inputSchema"]
            if "json" in input_schema:
                current_span.set_attribute("gen_ai.tool.json_schema", serialize(input_schema["json"]))

        invocation_state.update(
            {
                "agent": agent,
                "model": agent.model,
                "messages": agent.messages,
                "system_prompt": agent.system_prompt,
                "tool_config": ToolConfig(  # for backwards compatibility
                    tools=[{"toolSpec": tool_spec} for tool_spec in agent.tool_registry.get_all_tool_specs()],
                    toolChoice=cast(ToolChoice, {"auto": ToolChoiceAuto()}),
                ),
            }
        )

        # Retry loop for tool execution - hooks can set after_event.retry = True to retry
        while True:
            before_event, interrupts = await ToolExecutor._invoke_before_tool_call_hook(
                agent, tool_func, tool_use, invocation_state
            )

            if interrupts:
                yield ToolInterruptEvent(tool_use, interrupts)
                return

            if before_event.cancel_tool:
                cancel_message = (
                    before_event.cancel_tool if isinstance(before_event.cancel_tool, str) else "tool cancelled by user"
                )
                yield ToolCancelEvent(tool_use, cancel_message)

                cancel_result: ToolResult = {
                    "toolUseId": str(tool_use.get("toolUseId")),
                    "status": "error",
                    "content": [{"text": cancel_message}],
                }

                after_event, _ = await ToolExecutor._invoke_after_tool_call_hook(
                    agent,
                    None,
                    tool_use,
                    invocation_state,
                    cancel_result,
                    cancel_message=cancel_message,
                )
                yield ToolResultEvent(after_event.result)
                tool_results.append(after_event.result)
                return

            try:
                selected_tool = before_event.selected_tool
                tool_use = before_event.tool_use
                invocation_state = before_event.invocation_state

                if not selected_tool:
                    # Unknown tool: log here, but do NOT short-circuit. The middleware chain
                    # still runs with ctx.tool = None (matching TS), so middleware can observe
                    # or mock the call; the terminal produces the unknown-tool error result.
                    if tool_func == selected_tool:
                        logger.error(
                            "tool_name=<%s>, available_tools=<%s> | tool not found in registry",
                            tool_name,
                            list(agent.tool_registry.registry.keys()),
                        )
                    else:
                        logger.debug(
                            "tool_name=<%s>, tool_use_id=<%s> | a hook resulted in a non-existing tool call",
                            tool_name,
                            str(tool_use.get("toolUseId")),
                        )
                if structured_output_context.is_enabled:
                    kwargs["structured_output_context"] = structured_output_context

                # Run tool execution through the ExecuteToolStage middleware chain. The
                # terminal streams the tool and yields a plain ToolResultEvent as the last
                # (result) event; middleware can transform inputs/result, short-circuit with
                # a cached result, or gate execution behind an interrupt. A shallow copy of
                # tool_use guards its top-level keys (e.g. name, toolUseId) from accidental
                # in-place edits; its `input` can hold arbitrary, non-copyable objects (e.g.
                # the agent injected on direct tool calls) so it is shared by reference.
                # Middleware wanting an isolated tool_use should pass one via replace().
                middleware_context = ExecuteToolContext(
                    agent=agent,
                    tool=selected_tool,
                    tool_use=dict(tool_use),  # type: ignore[arg-type]
                    invocation_state=invocation_state,
                    _interrupt_state=agent._interrupt_state,
                )

                result_event: ToolResultEvent | None = None
                async for event in agent._middleware_registry.invoke(
                    ExecuteToolStage,
                    middleware_context,
                    _make_execute_tool_terminal(kwargs),
                ):
                    # Tool-originated interrupt: a ToolInterruptEvent yielded from tool.stream()
                    # (including sub-agent interrupts propagated via _AgentAsTool). Distinct from
                    # the middleware-initiated InterruptException handled below — this one rides
                    # the event stream rather than unwinding it. Register its interrupts so
                    # _interrupt_state.resume() can locate them by id, surface the event, and
                    # short-circuit here: a halted tool has no result, so the after-hook and the
                    # result handling below are intentionally skipped.
                    if isinstance(event, ToolInterruptEvent):
                        for interrupt in event.interrupts:
                            agent._interrupt_state.interrupts.setdefault(interrupt.id, interrupt)
                        yield event
                        return

                    # Capture the result but keep draining: middleware may yield trailing
                    # events after it, and the last ToolResultEvent wins (matching the model
                    # stage). It is re-emitted only after AfterToolCallEvent runs, since hooks
                    # may rewrite it. All non-result events flow through as they arrive.
                    if isinstance(event, ToolResultEvent):
                        result_event = event
                    else:
                        yield event

                if result_event is None:
                    raise RuntimeError(
                        "ExecuteToolStage middleware chain did not yield a ToolResultEvent. "
                        "Ensure middleware forwards events from next()."
                    )

                result = result_event.tool_result
                exception = result_event.exception

                after_event, _ = await ToolExecutor._invoke_after_tool_call_hook(
                    agent, selected_tool, tool_use, invocation_state, result, exception=exception
                )

                if ToolExecutor._should_retry(agent, after_event):
                    logger.debug("tool_name=<%s> | retry requested, retrying tool call", tool_name)
                    continue

                yield ToolResultEvent(after_event.result, exception=after_event.exception)
                tool_results.append(after_event.result)
                return

            except InterruptException as interrupt_exception:
                # Middleware-initiated interrupt (context.interrupt() with no response yet).
                # interrupt() is read-only, so this handler is the single place the interrupt
                # is registered before surfacing a ToolInterruptEvent to halt the agent,
                # matching how hook/tool interrupts are reported.
                agent._interrupt_state.interrupts.setdefault(
                    interrupt_exception.interrupt.id, interrupt_exception.interrupt
                )
                yield ToolInterruptEvent(tool_use, [interrupt_exception.interrupt])
                return

            except Exception as e:
                logger.exception("tool_name=<%s> | failed to process tool", tool_name)
                error_result: ToolResult = {
                    "toolUseId": str(tool_use.get("toolUseId")),
                    "status": "error",
                    "content": [{"text": f"Error: {str(e)}"}],
                }

                after_event, _ = await ToolExecutor._invoke_after_tool_call_hook(
                    agent, selected_tool, tool_use, invocation_state, error_result, exception=e
                )
                if ToolExecutor._should_retry(agent, after_event):
                    logger.debug("tool_name=<%s> | retry requested after exception, retrying tool call", tool_name)
                    continue
                yield ToolResultEvent(after_event.result, exception=after_event.exception)
                tool_results.append(after_event.result)
                return

    @staticmethod
    async def _stream_with_trace(
        agent: "Agent",
        tool_use: ToolUse,
        tool_results: list[ToolResult],
        cycle_trace: Trace,
        cycle_span: Any,
        invocation_state: dict[str, Any],
        structured_output_context: StructuredOutputContext | None = None,
        **kwargs: Any,
    ) -> AsyncGenerator[TypedEvent, None]:
        """Execute tool with tracing and metrics collection.

        Args:
            agent: The agent for which the tool is being executed.
            tool_use: Metadata and inputs for the tool to be executed.
            tool_results: List of tool results from each tool execution.
            cycle_trace: Trace object for the current event loop cycle.
            cycle_span: Span object for tracing the cycle.
            invocation_state: Context for the tool invocation.
            structured_output_context: Context for structured output management.
            **kwargs: Additional keyword arguments for future extensibility.

        Yields:
            Tool events with the last being the tool result.
        """
        tool_name = tool_use["name"]
        structured_output_context = structured_output_context or StructuredOutputContext()

        tracer = get_tracer()

        tool_call_span = tracer.start_tool_call_span(
            tool_use, cycle_span, custom_trace_attributes=agent.trace_attributes
        )
        tool_trace = Trace(f"Tool: {tool_name}", parent_id=cycle_trace.id, raw_name=tool_name)
        tool_start_time = time.time()

        with trace_api.use_span(tool_call_span):
            async for event in ToolExecutor._stream(
                agent, tool_use, tool_results, invocation_state, structured_output_context, **kwargs
            ):
                yield event

            if isinstance(event, ToolInterruptEvent):
                tool_duration = time.time() - tool_start_time
                if ToolExecutor._is_agent(agent):
                    agent.event_loop_metrics.add_tool_usage(tool_use, tool_duration, tool_trace, False)
                cycle_trace.add_child(tool_trace)
                tracer.end_tool_call_span(tool_call_span, tool_result=None)
                return

            result_event = cast(ToolResultEvent, event)
            result = result_event.tool_result

            tool_success = result.get("status") == "success"
            tool_duration = time.time() - tool_start_time
            message = Message(role="user", content=[{"toolResult": result}])
            if ToolExecutor._is_agent(agent):
                agent.event_loop_metrics.add_tool_usage(tool_use, tool_duration, tool_trace, tool_success, message)
            cycle_trace.add_child(tool_trace)

            tracer.end_tool_call_span(tool_call_span, result, error=result_event.exception)

    @abc.abstractmethod
    # pragma: no cover
    def _execute(
        self,
        agent: "Agent",
        tool_uses: list[ToolUse],
        tool_results: list[ToolResult],
        cycle_trace: Trace,
        cycle_span: Any,
        invocation_state: dict[str, Any],
        structured_output_context: "StructuredOutputContext | None" = None,
    ) -> AsyncGenerator[TypedEvent, None]:
        """Execute the given tools according to this executor's strategy.

        Args:
            agent: The agent for which tools are being executed.
            tool_uses: Metadata and inputs for the tools to be executed.
            tool_results: List of tool results from each tool execution.
            cycle_trace: Trace object for the current event loop cycle.
            cycle_span: Span object for tracing the cycle.
            invocation_state: Context for the tool invocation.
            structured_output_context: Context for structured output management.

        Yields:
            Events from the tool execution stream.
        """
        pass


def _make_execute_tool_terminal(
    extra_kwargs: dict[str, Any],
) -> "Any":
    """Build the terminal for the ExecuteToolStage middleware chain.

    The terminal streams the resolved tool and yields a plain ``ToolResultEvent`` as its
    last (result) event, matching the SDK-wide "last event is the result" convention.
    Intermediate ``ToolStreamEvent``s flow through unchanged. A tool-originated
    ``ToolInterruptEvent`` flows through as a normal event; the Output-phase adapter and
    the executor both recognize it as a control-flow signal rather than a result.

    A raw exception from ``tool.stream()`` is converted to an error ``ToolResultEvent`` here,
    inside the terminal, so ExecuteToolStage middleware always observes a result rather than a
    thrown exception (matching the TypeScript SDK). ``InterruptException`` is re-raised so a
    tool-raised interrupt still halts the agent instead of becoming an error result.

    All events are derived from ``ctx.tool_use`` (the possibly Input-transformed value the tool
    actually ran with), so the streamed, wrapped, and error events agree on identity fields
    (e.g. ``toolUseId``) even when an Input handler rewrote them.

    Args:
        extra_kwargs: Extra keyword arguments forwarded to ``tool.stream()``.

    Returns:
        An async generator function suitable as a middleware terminal.
    """

    async def terminal(ctx: ExecuteToolContext) -> AsyncGenerator[TypedEvent, None]:
        tool_use = ctx.tool_use

        # Unknown tool (not in the registry): the chain still ran so middleware could observe
        # or mock it, but with no tool to invoke the terminal yields the error result. The
        # message/exception mirror the pre-middleware unknown-tool contract.
        if ctx.tool is None:
            tool_name = tool_use["name"]
            yield ToolResultEvent(
                {
                    "toolUseId": str(tool_use.get("toolUseId")),
                    "status": "error",
                    "content": [{"text": f"Unknown tool: {tool_name}"}],
                },
                exception=Exception(f"Unknown tool: {tool_name}"),
            )
            return

        # Mirrors ToolExecutor._stream's original dispatch: built-in AgentTools yield
        # TypedEvents directly (ending in a ToolResultEvent); other tools yield raw values
        # we wrap in ToolStreamEvent, and their last raw value is the result.
        yielded_any = False
        last_raw_event: Any = None
        try:
            async for event in ctx.tool.stream(tool_use, ctx.invocation_state, **extra_kwargs):
                if isinstance(event, ToolInterruptEvent):
                    yield event
                    return

                if isinstance(event, ToolResultEvent):
                    # Re-emit so the exception decorated tools attach rides along as the result.
                    yield ToolResultEvent(event.tool_result, exception=event.exception)
                    return

                if isinstance(event, ToolStreamEvent):
                    yield event
                else:
                    yield ToolStreamEvent(tool_use, event)
                yielded_any = True
                last_raw_event = event
        except InterruptException:
            # A tool-raised interrupt must halt the agent — let it unwind rather than
            # becoming an error result (matches TS re-throwing InterruptError).
            raise
        except Exception as error:
            # Convert a raw tool failure to an error result inside the terminal so middleware
            # sees a result, not an exception. The executor's after-hook still receives the
            # exception via the ToolResultEvent below.
            logger.exception("tool_name=<%s> | tool execution failed", tool_use["name"])
            yield ToolResultEvent(
                {
                    "toolUseId": str(tool_use.get("toolUseId")),
                    "status": "error",
                    "content": [{"text": f"Error: {error}"}],
                },
                exception=error,
            )
            return

        # Non-SDK tool: no ToolResultEvent was emitted, so the last raw value is the result.
        # A tool that streamed nothing at all has no result — surface an error result rather
        # than a null one so the agent can continue (matches the pre-middleware degradation).
        if not yielded_any:
            yield ToolResultEvent(
                {
                    "toolUseId": str(tool_use.get("toolUseId")),
                    "status": "error",
                    "content": [{"text": f"Tool '{tool_use['name']}' did not return a result"}],
                }
            )
            return
        yield ToolResultEvent(cast(ToolResult, last_raw_event))

    return terminal
